src/libmongoc/src/mongoc/mongoc-cursor.c

```diff
@@ -662,11 +663,13 @@ _mongoc_cursor_fetch_stream (mongoc_cursor_t *cursor)
                                            &reply,
                                            &cursor->error);
    } else {
-      server_stream = mongoc_cluster_stream_for_reads (&cursor->client->cluster,
-                                                       cursor->read_prefs,
-                                                       cursor->client_session,
-                                                       &reply,
-                                                       &cursor->error);
+      server_stream =
```

Comment:
> TODO: This changeset works fully with recent servers, but when server selection falls back to the prior behavior of selecting a primary regardless of readPreference, the spec mandates that we must not send $readPreference with the aggregate command. This is surprisingly difficult given the current architecture of the code as these actions do not communicate with each other.

I am not sure if it sending `"$readPreference":{"mode":"secondary"}` in a command to a primary after overriding the read preference is an acceptable caveat. [This test](https://github.com/kevinAlbs/go-bootstrap/blob/a8a0bf43d9932f1c35c0ad4ff63f82c36f7279c7/investigations/sending_dollar_secondary_to_a_primary/main.go#L32-L37) sends a `ping` with secondary read preference. `"$readPreference":{"mode":"secondary"}` is in the command started events. The command succeeds in server versions back to 3.6.

cc @jmikola in case you are aware of other side-effects.

@vector-of-bool if this is not an acceptable caveat, would adding an out parameter, like `mongoc_read_prefs_t *effective_read_pref` to `_mongoc_cursor_fetch_stream` and passing that down to `mongoc_topology_description_suitable_servers` be a possible (though gross) solution?

Reply:
Even if an RS primary accepts `{ $readPreference: { mode: "secondary" }}`, that could be problematic if sent to a mongos or load balancer.

In https://github.com/mongodb/mongo-php-library/pull/876, I opted to check the list of servers in the topology first to decide whether or not to force a primary read preference. I imagine there's a more efficient way to do this in libmongoc (I had to resort to an extra server selection attempt if the topology was not already initialized), such as using an output parameter on a new server selection function.

Reply:
I put in the recommended changes to pass around the effective read mode and store it on the cursor and server stream. It's definitely not the prettiest solution, but it is the most un-intrusive one that I can think of.